### PR TITLE
[JIT] Metacompile boolean constants

### LINKED ISSUE #46703: [jit] metacompile if-statements with boolean literal or boolean constants
## Motivation

Skipping compilation when not needed is in general helpful for migrating code to torchscript.

## Pitch
```
import torch

class A(torch.nn.Module):
    good: torch.jit.Final[bool]
    def __init__(self):
        super().__init__()
        self.good = True
    def forward(self, x):
        #if hasattr(self, 'good'):  # works
        #if True:  # fails
        #if self.good is not None:  # works
        if self.good:  # fails
            return x

        # imagine below are some code that are scriptable depend on the availability of self.good
        print("hello") + 2  # shall not be compiled

x = A()
f = torch.jit.script(x)
f.forward(torch.randn(3))
```

cc @gmagogsfm

### PR DESCRIPTION
**Pull request by eellison**
Fixes https://github.com/pytorch/pytorch/issues/46703

Previously, we would compile one side of an if-statement if it was a type-based expression we could statically resolve. I think it's reasonable to extend this metacompilation to booleans that are constant at compile time. There have been some instances where i've recommended unintuitive workarounds due to not having this behavior. 

This is also possibly needed if we add boolean literals to schema declarations, which is a feature that might be needed to cleanup our `boolean_dispatch` mechanism.

### CODE REVIEW

**Review by gmagogsfm — COMMENTED** [review]
OOC, why do we need to handle this with specialized code rather than treating it as a normal case of const folding of `if`?

**Comment by eellison** [comment]
It's so that you can not compile code that wouldn't type check under one branch of the if.

File: torch/csrc/jit/frontend/ir_emitter.cpp [hunk 1]
```diff
@@ -1202,6 +1202,10 @@ struct to_ir {
         if (expr_out->node()->kind() == aten::is_scripting) {
           static_if = true;
         }
+        // MetaCompile on boolean literals and constants
```
Comment (ppwwyyxx) [inline torch/csrc/jit/frontend/ir_emitter.cpp:1205]: line169 has some comment that appears related and confusing:
```
  c10::optional<bool>
      static_if_; // certain expression cause us to emit a static if statement
                  // this value is present if this is the case.
                  // this is not equivalent to value_ being a constant
                  // it is possible for value_ to be constant but for
                  // the expression that produced it to not trigger the
                  // static if behavior. e.g. use of a variable assigned
                  // to a constant
```
I don't know if they need to be updated to reflect this PR

**Review by SplitInfinity — COMMENTED** [review]
Very cool! A few questions:

1. Should we add a test with `__constants__`? It's a deprecated feature but this could benefit old code that still uses it.
2. What about cases in which the metacompiled boolean is inside a constant list, dict, tuple or class type attribute?
3. Two of the test failures are code that is some variant of
```
            x = []  # Inferred to be a list of Tensors
            if True:
                x = [1, 2, 3]  # List of ints
```
that compiles now but didn't before. I guess this isn't much of a question but I think this code should still fail to compile.

**Comment by eellison** [comment]
> Very cool! A few questions:
> 
> 1. Should we add a test with `__constants__`? It's a deprecated feature but this could benefit old code that still uses it.
> 2. What about cases in which the metacompiled boolean is inside a constant list, dict, tuple or class type attribute?
> 3. Two of the test failures are code that is some variant of
> 
> ```
>             x = []  # Inferred to be a list of Tensors
>             if True:
>                 x = [1, 2, 3]  # List of ints
> ```
> 
> that compiles now but didn't before. I guess this isn't much of a question but I think this code should still fail to compile.

The `torch.jit.Final` is equivalent to `__constants__`.  I don't know what you mean metacompiled boolean is inside a constant list, dict, tuple or class type attribute? The boolean is still a boolean. 

For 3 - the point of this PR is to not unnecessarily compile things that will never get run. If we want that example to not compile then we shouldn't accept this. I don't think users really do things like "if True" and "if False", so the cases that this will actually be a literal value is exclusively for module constants (and in the future, literal value arguments). If we're going to allow users to metacompile with type expressions, I think we should also allow them to use booleans, as they're more intuitive. 

We can also defer this PR until the literal's is implemented, when it will be more clearly needed.

**Comment by SplitInfinity** [comment]
> The torch.jit.Final is equivalent to __constants__. I don't know what you mean metacompiled boolean is inside a constant list, dict, tuple or class type attribute? The boolean is still a boolean.

I mean that the boolean used in the `if` is a member of a some container type that is constant. I don't doubt that it will work, I just think it would be good to have a test case for it.

>  I don't think users really do things like "if True" and "if False"

True but I think the example I gave could have been written with a model attribute constant too (e.g. `self.val` instead of `True`).

**Review by sab-regime — APPROVED** [review]

**Comment by ppwwyyxx** [comment]
any updates?

**Comment by eellison** [comment]
I think we should land this just waiting on a response from @gmagogsfm or @SplitInfinity. I dont see why we should allow:
```
if self.val is not None:
    "abc" + 1
```
where `val` is known to be None, but not allow:

```
if self.val
    "abc" + 1
```
where self.val is known to be false. The former is relying on type deduction, where as the latter is much more explicit. It is easy to construct a wonky `if False` example, but no one is actually doing this. These are only constants if a user had opted into them with `torch.jit.Final`.  The same wonky examples could also be constructed with `if None is not None` today. 

I had a user ask to do something similar with metaprogramming of the forward and they ended up using htis: https://gist.github.com/eellison/4e0ded4f390c2af61a380fd51384bd9a . I think that is strictly more confusing than if they had been able to replace `isBar` with a tuple containing true and false.

**Review by SplitInfinity — APPROVED** [review]
Yeah, I guess my examples were academic. I also didn't know until two days ago that `__annotations__ = {'a': int}` is equivalent to 
```
class A:
   a: int
```

I think you need to fix up some test failures but looks good otherwise.

**Comment by ppwwyyxx** [comment]
Would love to see this landed :rocket:

**Comment by eellison** [comment]
@ppwwyyxx sure, sorry for delay, will try to land it today

**Pull request closed without merging**